dev/AppLifecycle/ActivationRegistrationManager.cpp

```diff
@@ -8,47 +8,89 @@
 #include "ProtocolActivatedEventArgs.h"
 #include "FileActivatedEventArgs.h"
 #include "Association.h"
+#include "ExtensionContract.h"
+#include "EncodedLaunchExecuteCommand.h"
+
+extern "C" HRESULT __stdcall DllRegisterServer() noexcept;
 
 namespace winrt::Microsoft::ProjectReunion::implementation
 {
+    std::wstring GenerateEncodedLaunchUri(std::wstring const& appUserModelId, std::wstring const& contractId)
+    {
+        return c_encodedLaunchSchemeName + L"://" + appUserModelId + L"?ContractId=" + contractId;
+    }
+
+    std::wstring GenerateCommandLine(std::wstring const& modulePath)
+    {
+        std::wstring command = modulePath;
+
+        if (modulePath.empty())
+        {
+            command = GetModulePath();
+        }
+
+        command += L" " + c_argumentPrefix + c_protocolArgumentString + c_argumentSuffix;
+        return command;
+    }
+
     void ActivationRegistrationManager::RegisterForFileTypeActivation(
-        array_view<hstring const> supportedFileTypes, array_view<hstring const> supportedVerbs,
-        hstring const& applicationDisplayName, hstring const& logo)
+        array_view<hstring const> supportedFileTypes, hstring const& logo, hstring const& displayName,
+        array_view<hstring const> supportedVerbs, hstring const& exePath)
     {
         if (HasIdentity())
         {
             throw hresult_illegal_method_call();
         }
 
-        auto appId = ComputeAppId();
+        auto appId = ComputeAppId(exePath.c_str());
         auto type = AssociationType::File;
         auto progId = ComputeProgId(appId, type);
 
         if (supportedFileTypes.size())
         {
-            RegisterProgId(progId.c_str(), L"", applicationDisplayName.c_str(), logo.c_str());
+            RegisterProgId(progId.c_str(), L"", displayName.c_str(), logo.c_str());
             RegisterApplication(appId.c_str());
         }
 
-        for (auto extension : supportedFileTypes)
+        for (auto fileType : supportedFileTypes)
         {
-            RegisterFileExtension(extension.c_str());
+            RegisterFileExtension(fileType.c_str());
             
             for (auto verb : supportedVerbs)
             {
-                auto command = GetModulePath();
-                command += L" ----" + c_fileArgumentString + L":" + verb.c_str() + L",%1";
+                auto command = GenerateCommandLine(exePath.c_str()) + GenerateEncodedLaunchUri(L"App",
+                    c_fileContractId) + L"&Verb=" + verb.c_str() + L"&File=%1";
```

Comment:
>verb [](start = 52, length = 4)

Any restrictions on allowed values? Required to be ascii? Can contain punctuation % & \ / : yadda yadda or is that invalid? I suspect implementation dependent errors can be returned e.g. if verb = "Cookies&Cream" or "A\B".

Bad input should result in E_INVALIDARG

Reply:
Not really as this is a process command line with arguments.  It is parsed by ShellExecute and a set of platform APIs before execution.  We will not be providing any preventative validation unless we find evidence of it being a common error to get it wrong, and why.